lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -723,6 +723,17 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitness(ValueDecl *req,
       return true;
     }
 
+    // Allow match of dynamic Self to the conforming type, if it's
+    // a non-class or final class.
+    bool mismatch(DynamicSelfType *firstType, TypeBase *secondType,
+                  Type sugaredFirstType) {
+      if (secondType->isEqual(Conformance->getType())) {
+        auto classDecl = Conformance->getType()->getClassOrBoundGenericClass();
+        return !classDecl || classDecl->isFinal();
```

Comment:
Yes, `firstType` is from the protocol requirement, so it is dynamic self + generic param 0,0. But because we don't actually have the protocol conformance done, there isn't yet a generic environment that will map that param to the (hopefully) conforming type.

So I think this fake way of directly substituting the conforming type to compare is the only way to go here?

Reply:
What I'm suggesting is changing matchTypes() in inferTypeWitnessesViaValueWitness() to strip off the dynamic self type by calling eraseDynamicSelfType() on the requirement type first. Then you don't need this case at all.

Reply:
Took me awhile, but I finally figured out what you meant. This is cleaner.